libtock/alarm_timer.c

```diff
@@ -215,33 +215,37 @@ int yield_for_with_timeout(bool* cond, uint32_t ms) {
 int gettimeasticks(struct timeval *tv, __attribute__ ((unused)) void *tzvp)
 {
   uint32_t frequency, now, seconds, remainder;
-
   alarm_internal_frequency(&frequency);
   alarm_internal_read(&now);
 
-  // The microsecond calculation will overflow in the intermediate scaling of
-  // (remainder * 1000) if the remainder is approximately greater than 4e6. Because
-  // remainder is calculated as now % frequency, we can define 0 <= remainder < frequency.
-  // This implies that the tv_usec must be of type uint64_t if frequency > 4MHz to avoid
-  // an overflow from occurring. We check this in the below assertion statement.
-  const uint32_t MAX_VALID_FREQ = 4000000;
-  assert(frequency < MAX_VALID_FREQ || sizeof(tv->tv_usec) == sizeof(uint64_t));
-
-  // Confirm frequency assumption
-  assert(frequency > 0);
+  // Confirm frequency assumptions. Since remainder is
+  // ticks % frequency, remainder <= frequency. This assertion
+  // guards against overflowing the us calculation. At worst,
+  // remainder == frequency and scaling remainder to convert
+  // to us will overflow. We confirm the greatest possible
+  // remainder value (frequency) scaled by 1e6 will fit within
+  // a uint64 and guard against division by zero.
+  assert(frequency > 0 && (frequency * 1e6) <= UINT64_MAX);
```

Comment:
`frequency` is still a 32-bit value, so won't it always be less then `UINT64_MAX`?

Reply:
```suggestion
  assert(frequency > 0ull && (frequency * 1e6ull) <= UINT64_MAX);
```

I believe Alistair is correct here, though you can [make the constants 64-bit types](https://stackoverflow.com/questions/2835746/how-do-you-specify-a-64-bit-unsigned-int-const-0x8000000000000000-in-vs2008), and [standard type promotion](https://c0x.shape-of-code.com/6.3.1.8.html) should do the right thing from there.

Reply:
I think we can remove the overflow check given this will never overflow (so long as the convention remains of frequency being a `uint32_t`

Reply:
```
#if __has_builtin(__builtin_types_compatible_p)
static_assert(__builtin_types_compatible_p(uint32_t, typeof(frequency))
#endif
```

You can always [enforce that assumption](https://stackoverflow.com/a/55804363) :)